source/falaise/snemo/datamodels/sim_tracker_digi_hit.cc

```diff
@@ -5,239 +5,321 @@
-  this->geomtools::base_hit::tree_dump(out_, title_, indent_, true);
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Anode t0 : ";
-  if (has_anode_t0()) {
-    out_ << _anode_t0_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Anode t1 : ";
-  if (has_anode_t1()) {
-    out_ << _anode_t1_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Anode t2 : ";
-  if (has_anode_t2()) {
-    out_ << _anode_t2_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Anode t3 : ";
-  if (has_anode_t3()) {
-    out_ << _anode_t3_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Anode t4 : ";
-  if (has_anode_t4()) {
-    out_ << _anode_t4_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Cathode t5 : ";
-  if (has_cathode_t5()) {
-    out_ << _cathode_t5_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::tag << "Cathode t6 : ";
-  if (has_cathode_t6()) {
-    out_ << _cathode_t6_;
-  } else {
-    out_ << "<none>";
-  }
-  out_ << std::endl;
-
-  out_ << indent_ << datatools::i_tree_dumpable::inherit_tag(inherit_)
-       << "Validity : " << (is_valid() ? "<yes>" : "<no>") << std::endl;
-
-  return;
-}
-
-}  // end of namespace datamodel
+  namespace datamodel {
+
+    // Serial tag :
+    DATATOOLS_SERIALIZATION_SERIAL_TAG_IMPLEMENTATION(sim_tracker_digi_hit,
+						      "snemo::datamodel::sim_tracker_digi_hit")
+
+    sim_tracker_digi_hit::sim_tracker_digi_hit()
+    {
+      return;
+    }
+
+    sim_tracker_digi_hit::~sim_tracker_digi_hit()
+    {
+      return;
+    }
+
+    bool sim_tracker_digi_hit::is_valid() const
+    {
+      if (!has_hit_id())
+	{
+	  return false;
+	}
+      if (!has_geom_id())
+	{
+	  return false;
+	}
+      if (!has_anode_R0())
+	{
+	  return false;
+	}
+      return true;
+    }
+
+    void sim_tracker_digi_hit::reset()
+    {
+      reset_anode_R0();
+      reset_anode_R1();
+      reset_anode_R2();
+      reset_anode_R3();
+      reset_anode_R4();
+      reset_cathode_R5();
+      reset_cathode_R6();
+      return;
+    }
+
+    bool sim_tracker_digi_hit::has_elec_id() const
+    {
+      return _elec_id_.is_valid();
+    }
+
+    void sim_tracker_digi_hit::invalidate_elec_id()
+    {
+      _elec_id_.invalidate();
+      return;
+    }
+
+    const geomtools::geom_id & sim_tracker_digi_hit::get_elec_id() const
+    {
+      return _elec_id_;
+    }
+
+    geomtools::geom_id & sim_tracker_digi_hit::grab_elec_id()
+    {
+      return _elec_id_;
+    }
+
+    void sim_tracker_digi_hit::set_elec_id(const geomtools::geom_id & a_gid)
+    {
+      _elec_id_ = a_gid;
+      return;
+    }
+
```

Comment:
Again, surely an electronic id is really an invariant of the instance once created? Why would it change *after* the data is created?

Reply:
I think I want to keep this interface for EID because a `sim_tracker_digi_hit` can be created with a given GID and then the EID will be set once the GID has been converted into an EID by the electronic mapping.
A `sim_tracker_digi_hit` and a `sim_calo_digi_hit` have 2 `geomtools::geom_id` : 
- A Geometric ID (cell ID with Module/Row/Layer etc... for example)
- And an Electronic ID (channel ID with : Crate:Board:Feast:Channel)
It is useful to have these 2 IDs during the digi process and maybe it will be useful later so we store them both